File: leetcode-sorting-algorithms-solutions-advanced/sorting_problems/group_anagrams.py
from collections import defaultdict

# Problem: Group Anagrams
# Given an array of strings `strs`, group the anagrams together.
# You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
# typically using all the original letters exactly once.

# --- Approach 1: Sort Characters to Create a Key ---
# Time Complexity: O(N * L log L)
#   N: number of strings in `strs`
#   L: maximum length of a string
#   For each string, we sort its characters, which takes O(L log L).
#   We do this for N strings. Dictionary operations (insertion/lookup) are O(L) on average
#   because the key is a string of length L.
# Space Complexity: O(N * L) in the worst case, where all strings are unique anagrams,
#   and we store all N strings and their sorted keys.

def group_anagrams_sort_key(strs: list[str]) -> list[list[str]]:
    """
    Groups anagrams by sorting the characters of each word to create a canonical key.

    Args:
        strs (list[str]): A list of strings.

    Returns:
        list[list[str]]: A list of lists, where each inner list contains anagrams.
    """
    if not strs:
        return []

    # Use a dictionary where keys are sorted tuples of characters (canonical form)
    # and values are lists of original strings that are anagrams.
    anagram_groups = defaultdict(list)

    for s in strs:
        # Convert string to a list of characters, sort it, then convert back to a tuple.
        # A tuple is used as a dictionary key because lists are mutable and cannot be keys.
        # A string could also be used: `key = "".join(sorted(s))`
        sorted_s_tuple = tuple(sorted(s))
        anagram_groups[sorted_s_tuple].append(s)

    # Convert the dictionary values (lists of anagrams) into a list of lists.
    return list(anagram_groups.values())

# Example: strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
# "eat" -> sorted("eat") -> ('a', 'e', 't') -> anagram_groups[('a', 'e', 't')] = ["eat"]
# "tea" -> sorted("tea") -> ('a', 'e', 't') -> anagram_groups[('a', 'e', 't')] = ["eat", "tea"]
# "tan" -> sorted("tan") -> ('a', 'n', 't') -> anagram_groups[('a', 'n', 't')] = ["tan"]
# "ate" -> sorted("ate") -> ('a', 'e', 't') -> anagram_groups[('a', 'e', 't')] = ["eat", "tea", "ate"]
# "nat" -> sorted("nat") -> ('a', 'n', 't') -> anagram_groups[('a', 'n', 't')] = ["tan", "nat"]
# "bat" -> sorted("bat") -> ('a', 'b', 't') -> anagram_groups[('a', 'b', 't')] = ["bat"]
# Result: [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]


# --- Approach 2: Count Characters to Create a Key (More efficient for fixed, small alphabet) ---
# Time Complexity: O(N * (L + K)) where K is the size of the alphabet (e.g., 26 for lowercase English).
#   N: number of strings in `strs`
#   L: maximum length of a string
#   For each string, we iterate through its characters to build a count array, which takes O(L).
#   Then we convert this count array (or tuple) to a key, which takes O(K) time.
#   Total for N strings: O(N * (L + K)).
#   This is better than O(N * L log L) if K is significantly smaller than L log L.
#   For English alphabet (K=26), this is effectively O(N * L) as K is a constant.
# Space Complexity: O(N * L) in the worst case (similar to sort_key approach).

def group_anagrams_count_key(strs: list[str]) -> list[list[str]]:
    """
    Groups anagrams by counting character frequencies to create a canonical key.
    This approach is particularly efficient for a fixed, small alphabet like lowercase English letters.

    Args:
        strs (list[str]): A list of strings.

    Returns:
        list[list[str]]: A list of lists, where each inner list contains anagrams.
    """
    if not strs:
        return []

    anagram_groups = defaultdict(list)

    for s in strs:
        # Create a count array (or tuple) for the characters in the string.
        # For lowercase English letters, an array of 26 zeros is sufficient.
        # The index represents 'a' + offset (e.g., 0 for 'a', 1 for 'b', ..., 25 for 'z').
        char_counts = [0] * 26
        for char in s:
            char_counts[ord(char) - ord('a')] += 1

        # Convert the count array to a tuple to use as a dictionary key.
        # Example: "aab" -> (2,1,0,0,...) for ('a','b',...)
        count_tuple_key = tuple(char_counts)
        anagram_groups[count_tuple_key].append(s)

    return list(anagram_groups.values())

# Example: strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
# "eat" -> counts[0]=1, counts[4]=1, counts[19]=1 -> (1,0,0,0,1,0,...,0,1,0,0,0,0) -> anagram_groups[...] = ["eat"]
# "tea" -> counts[0]=1, counts[4]=1, counts[19]=1 -> (1,0,0,0,1,0,...,0,1,0,0,0,0) -> anagram_groups[...] = ["eat", "tea"]
# ... and so on.

# Comparing the two approaches:
# - `sort_key`: More general, works for any character set (Unicode, numbers, symbols) without modification.
#   Its performance depends on the efficiency of the sorting algorithm for strings/lists of characters.
# - `count_key`: Optimized for fixed, small alphabets. If the alphabet is very large or unknown,
#   this approach might require a hash map for character counts which would then have different performance
#   characteristics (O(L) to build map, O(K') to convert to tuple where K' is number of unique chars).
#   For standard interview settings (lowercase English), `count_key` is often preferred for its
#   better theoretical time complexity.

File: leetcode-hash-tables-solutions-pro/src/problems/groupAnagrams.ts
```typescript
/**
 * Given an array of strings `strs`, group the anagrams together.
 * You can return the answer in any order.
 *
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
 * typically using all the original letters exactly once.
 *
 * Example 1:
 * Input: strs = ["eat","tea","tan","ate","nat","bat"]
 * Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
 *
 * Example 2:
 * Input: strs = [""]
 * Output: [[""]]
 *
 * Example 3:
 * Input: strs = ["a"]
 * Output: [["a"]]
 *
 * Constraints:
 * 1 <= strs.length <= 10^4
 * 0 <= strs[i].length <= 100
 * strs[i] consists of lowercase English letters.
 */

/**
 * Optimal Solution: Using a Hash Map (Map in TypeScript/JavaScript)
 *
 * Approach:
 * The core idea is that anagrams will have the same "canonical" representation.
 * If we can transform each string into such a canonical form, we can use this
 * form as a key in a hash map to group all strings that share that form.
 *
 * 1. Initialize an empty hash map where:
 *    - Key: The canonical form of a string (e.g., sorted string, or a frequency count string).
 *    - Value: A list of strings that are anagrams and share this canonical form.
 * 2. Iterate through each `str` in the input array `strs`.
 * 3. For each `str`, generate its canonical key:
 *    a. **Method 1 (Sorted String - chosen for implementation):** Convert the string to a character array, sort it, and then join it back into a string.
 *       Example: "tea" -> ['t','e','a'] -> ['a','e','t'] -> "aet"
 *                "eat" -> ['e','a','t'] -> ['a','e','t'] -> "aet"
 *    b. **Method 2 (Frequency Count String):** Create a frequency array (e.g., of size 26 for lowercase English letters). Count character occurrences. Convert this frequency array into a unique string key.
 *       Example: "tea" -> [1,0,0,0,1,0,...1...] (counts for a,b,c,d,e,f,...t...) -> "1#0#0#1#0#1" (delimiter helps)
 *       This method can sometimes be faster if string sorting is expensive for very long strings, as it's O(M) where M is string length,
 *       while sorting is O(M log M). However, M is usually small in constraints.
 * 4. Use this `key` to interact with the hash map:
 *    a. If `map.has(key)` is false, create a new list for this key: `map.set(key, [])`.
 *    b. Add the original `str` to the list associated with `key`: `map.get(key)!.push(str)`.
 * 5. After iterating through all strings, the values of the hash map will be lists of grouped anagrams.
 *    Return `Array.from(map.values())`.
 *
 * Time Complexity: O(N * M log M) or O(N * M)
 * - N is the number of strings in `strs`.
 * - M is the average length of a string in `strs`.
 * - Generating the key:
 *   - If sorting strings: O(M log M) for each string. Total: O(N * M log M).
 *   - If using frequency count array: O(M) for each string. Converting array to string can be O(alphabet_size) which is constant (26). Total: O(N * M).
 * - Hash map operations (insert/lookup): O(1) on average for a string key of length M.
 *
 * Space Complexity: O(N * M)
 * - In the worst case, all strings could be unique anagram groups. The map would store N keys, and the lists would collectively store all characters of all N strings.
 * - The sorted string key also takes O(M) space temporarily for each string.
 */
export function groupAnagrams(strs: string[]): string[][] {
    // A Map to store grouped anagrams.
    // The key will be the canonical form (sorted string), and the value will be an array of anagrams.
    const anagramGroups = new Map<string, string[]>();

    for (const str of strs) {
        // 3. Generate the canonical key for the current string.
        // Convert to character array, sort it, and join back to a string.
        // Example: "eat" -> ['e','a','t'] -> ['a','e','t'] -> "aet"
        const key = str.split('').sort().join('');

        // 4. Group strings using the generated key.
        // If the key doesn't exist, initialize an empty array for it.
        if (!anagramGroups.has(key)) {
            anagramGroups.set(key, []);
        }
        // Add the original string to the list associated with its canonical key.
        anagramGroups.get(key)!.push(str);
    }

    // 5. Return all the grouped lists of anagrams.
    // `Array.from(map.values())` extracts all value arrays into a new array.
    return Array.from(anagramGroups.values());
}
```